go/parquet/metadata/file.go

```diff
@@ -0,0 +1,478 @@
+}
+
+// NewFileMetadataBuilder will use the default writer properties if nil is passed for
+// the writer properties and nil is allowable for the key value metadata.
+func NewFileMetadataBuilder(schema *schema.Schema, props *parquet.WriterProperties, kvmeta KeyValueMetadata) *FileMetaDataBuilder {
+	var crypto *format.FileCryptoMetaData
+	if props.FileEncryptionProperties() != nil && props.FileEncryptionProperties().EncryptedFooter() {
+		crypto = format.NewFileCryptoMetaData()
+	}
+	return &FileMetaDataBuilder{
+		metadata:       format.NewFileMetaData(),
+		props:          props,
+		schema:         schema,
+		kvmeta:         kvmeta,
+		cryptoMetadata: crypto,
+	}
+}
+
+// GetFileCryptoMetaData returns the cryptographic information for encrypting/
+// decrypting the file.
+func (f *FileMetaDataBuilder) GetFileCryptoMetaData() *FileCryptoMetadata {
+	if f.cryptoMetadata == nil {
+		return nil
+	}
+
+	props := f.props.FileEncryptionProperties()
+	f.cryptoMetadata.EncryptionAlgorithm = props.Algorithm().ToThrift()
+	keyMetadata := props.FooterKeyMetadata()
+	if keyMetadata != "" {
+		f.cryptoMetadata.KeyMetadata = []byte(keyMetadata)
+	}
+
+	return &FileCryptoMetadata{f.cryptoMetadata, 0}
+}
+
+// AppendRowGroup adds a rowgroup to the list and returns a builder
+// for that row group
+func (f *FileMetaDataBuilder) AppendRowGroup() *RowGroupMetaDataBuilder {
+	if f.rowGroups == nil {
+		f.rowGroups = make([]*format.RowGroup, 0, 1)
+	}
+
+	rg := format.NewRowGroup()
+	f.rowGroups = append(f.rowGroups, rg)
+	f.currentRgBldr = NewRowGroupMetaDataBuilder(f.props, f.schema, rg)
+	return f.currentRgBldr
+}
+
+// Finish will finalize the metadata of the number of rows, row groups,
+// version etc. This will clear out this filemetadatabuilder so it can
+// be re-used
+func (f *FileMetaDataBuilder) Finish() (*FileMetaData, error) {
+	totalRows := int64(0)
+	for _, rg := range f.rowGroups {
+		totalRows += rg.NumRows
+	}
+	f.metadata.NumRows = totalRows
+	f.metadata.RowGroups = f.rowGroups
+	switch f.props.Version() {
+	case parquet.V1:
+		f.metadata.Version = 1
+	case parquet.V2:
+		f.metadata.Version = 2
+	default:
+		f.metadata.Version = 0
+	}
+	createdBy := f.props.CreatedBy()
+	f.metadata.CreatedBy = &createdBy
+
+	// Users cannot set the `ColumnOrder` since we donot not have user defined sort order
+	// in the spec yet.
+	// We always default to `TYPE_DEFINED_ORDER`. We can expose it in
+	// the API once we have user defined sort orders in the Parquet format.
+	// TypeDefinedOrder implies choose SortOrder based on ConvertedType/PhysicalType
+	typeDefined := format.NewTypeDefinedOrder()
+	colOrder := &format.ColumnOrder{TYPE_ORDER: typeDefined}
+	f.metadata.ColumnOrders = make([]*format.ColumnOrder, f.schema.NumColumns())
+	for idx := range f.metadata.ColumnOrders {
+		f.metadata.ColumnOrders[idx] = colOrder
+	}
+
+	fileEncProps := f.props.FileEncryptionProperties()
+	if fileEncProps != nil && !fileEncProps.EncryptedFooter() {
```

Comment:
is there a reason to allow nil fileEncProps (can defaults always indicate no encryption?)

Reply:
so, Go doesn't allow default values for function params like C++. The standard pattern would be to have a separate function for the case with encryption or not or to use the options pattern. I'll take a look and see which pattern i like better for this.

Reply:
Ah, i misunderstood the question here. 

The defaults are that the `FileEncryptionProperties()` are nil which indicates no encryption.